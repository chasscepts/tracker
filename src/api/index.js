import axios from 'axios';

const normalizeError = (err) => {
  if (!err) {
    return { message: 'An unknown error encountered. Please try again.' };
  }
  if (err.response) {
    return { message: err.response.data.message || JSON.stringify(err.response.data) };
  }
  if (err.request) {
    return { message: 'Server is not responding.' };
  }
  if (err.message) {
    return { message: err.message };
  }
  if (typeof err === 'string') {
    return { message: err };
  }
  return { message: 'An unknown error encountered. Please try again.' };
};

const instantiate = (token) => {
  const instance = axios.create({
    baseURL: 'http://localhost:3000/',
    responseType: 'json',
  });
  instance.defaults.headers.common.Authorization = `Bearer ${token}`;

  return instance;
};

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const get = (instance, path) => new Promise((resolve, reject) => {
  instance.get(path)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @param {{ string: any }} data sent in body of this post
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const post = (instance, path, data) => new Promise((resolve, reject) => {
  instance.post(path, data)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

const api = {
  getGroups: (token) => get(instantiate(token), '/groups'),
  getTasks: (token, start, end) => {
    let path = '/tasks';
    if (start) {
      path = `${path}?start=${start}`;
      if (end) {
        path = `${path}&end=${end}`;
      }
    }

    return get(instantiate(token), path);
  },
  updateTask: (token, title) => post(instantiate(token), '/tasks', { title }),
};

export default api;
