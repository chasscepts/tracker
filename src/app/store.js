import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../reducers/authSlice';
import tasksReducer from '../reducers/tasksSlice';
import timerReducer from '../reducers/timerSlice';

export default configureStore({
  reducer: {
    auth: authReducer,
    tasks: tasksReducer,
    timer: timerReducer,
  },
});
