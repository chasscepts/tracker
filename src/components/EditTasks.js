import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { COLORS } from '../utilities';
import { today } from '../utilities/dates';
import button from '../assets/css/button.module.css';
import {
  selectUpdateTaskRequestCount,
  selectTasks,
  loadTasks,
  deleteTaskAsync,
  updateTaskAsync,
} from '../reducers/tasksSlice';
import LdsRing from './LdsRing';
import LoadingPanel from './LoadingPanel';

const styles = {
  container: {
    height: '100%',
    overflow: 'hidden',
    display: 'flex',
    flexDirection: 'column',
  },
  form: {
    width: '100%',
    maxWidth: '400px',
    border: '1px solid #ddd',
    borderRadius: '5px',
    padding: '15px',
    margin: '5px auto 15px',
  },
  h2: {
    marginBottom: '15px',
    color: COLORS.primary,
    borderBottom: `1px dotted ${COLORS.primary}`,
    paddingBottom: '5px',
  },
  input: {
    display: 'block',
    width: '100%',
    padding: '7px',
    border: '1px solid #ddd',
    marginBottom: '15px',
    color: 'inherit',
  },
  controls: {
    display: 'flex',
    justifyContent: 'flex-end',
  },
  deleteBtn: {
    marginLeft: '20px',
  },
  loadingWrap: {
    padding: '5px',
    display: 'flex',
    alignItems: 'center',
    border: '1px solid #eee',
    boxShadow: '0 0 1px 1px',
    borderRadius: '5px',
    marginBottom: '15px',
  },
  loadingMsg: {
    marginLeft: '10px',
  },
  error: {
    position: 'relative',
    fontWeight: 'bold',
    fontSize: '0.8rem',
    padding: '8px',
    color: COLORS.red,
    border: '1px solid #ddd',
    textAlign: 'center',
    marginBottom: '15px',
    maxHeight: '150px',
    overflow: 'auto',
  },
  errorBtn: {
    position: 'absolute',
    right: '3px',
    top: '3px',
    padding: 0,
    border: 'none',
    color: COLORS.red,
    backgroundColor: 'transparent',
    cursor: 'pointer',
  },
  tasksPanelWrap: {
    flex: 1,
    overflow: 'auto',
  },
  tasksPanel: {
    display: 'flex',
    justifyContent: 'space-between',
    flexFlow: 'row wrap',
    padding: '10px',
  },
  taskBtn: {
    width: '80px',
    height: '80px',
    padding: '5px',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'transparent',
    border: '1px solid #ddd',
    borderRadius: '4px',
    margin: '10px',
    cursor: 'pointer',
    color: 'inherit',
    fontWeight: 'bold',
    fontSize: '0.9rem',
  },
  activeTaskBtn: {
    backgroundColor: COLORS.primary,
    color: '#fff',
  },
  taskBtnText: {
    pointerEvents: 'none',
  },
  selectedTask: {
    fontWeight: 'bold',
    fontSize: '0.8rem',
    marginBottom: '5px',
  },
  noSelectedTask: {
    fontWeight: 'bold',
    fontSize: '0.8rem',
    marginBottom: '5px',
    color: COLORS.red,
  },
  selectedTaskLabel: {
    color: COLORS.primary,
  },
};

export default function EditTask() {
  const [title, setTitle] = useState('');
  const [selectedTask, selectTask] = useState(null);
  const requestCount = useSelector(selectUpdateTaskRequestCount);
  const tasks = useSelector(selectTasks);
  const dispatch = useDispatch();

  if (!tasks) {
    dispatch(loadTasks(today()));
  }

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    if (name === 'title') {
      setTitle(value);
    }
  };

  const submitForm = () => {
    if (title && selectedTask) {
      dispatch(updateTaskAsync(selectedTask.id, title));
    }
  };

  const handleEnterKey = (evt) => {
    if (evt.key === 'Enter') {
      submitForm();
    }
  };

  const removeTask = () => {
    if (selectedTask) {
      dispatch(deleteTaskAsync(selectedTask.id));
      selectTask(null);
    }
  };

  const handleTaskClick = (evt) => {
    const { name } = evt.target;
    const task = tasks.find((t) => t.id === +name);
    selectTask(task);
    if (task) {
      setTitle(task.title);
    }
  };

  const btnStyle = (id) => {
    if (selectedTask && selectedTask.id === id) {
      return { ...styles.taskBtn, ...styles.activeTaskBtn };
    }
    return styles.taskBtn;
  };

  const loadingMessage = requestCount > 1
    ? `${requestCount} requests ongoing` : `${requestCount} request ongoing`;

  return (
    <div style={styles.container}>
      <div style={styles.form}>
        <h2 style={styles.h2}>Edit Task</h2>
        {requestCount > 0
        && (
        <div style={styles.loadingWrap}>
          <LdsRing width={20} color={COLORS.primary} />
          <div style={styles.loadingMsg}>{loadingMessage}</div>
        </div>
        )}
        {selectedTask && (
        <div style={styles.selectedTask}>
          <span style={styles.selectedTaskLabel}>Selected Task: </span>
          <span>{selectedTask.title}</span>
        </div>
        )}
        {!selectedTask && (
        <div style={styles.noSelectedTask}>You have not selected any task</div>
        )}
        <input
          style={styles.input}
          name="title"
          value={title}
          onChange={handleChange}
          onKeyDown={handleEnterKey}
          placeholder="Enter New Title"
        />
        <div style={styles.controls}>
          <button
            className={`${button.btn} ${button.blue} ${button.round}`}
            type="button"
            onClick={submitForm}
          >
            Edit
          </button>
          <button
            style={styles.deleteBtn}
            className={`${button.btn} ${button.red} ${button.round}`}
            type="button"
            onClick={removeTask}
          >
            Delete
          </button>
        </div>
      </div>
      {tasks && (
      <div style={styles.tasksPanelWrap}>
        <div style={styles.tasksPanel}>
          {tasks.map((t) => (
            <button
              style={btnStyle(t.id)}
              key={t.title}
              name={t.id}
              type="button"
              onClick={handleTaskClick}
            >
              <span style={styles.taskBtnText}>{t.title}</span>
            </button>
          ))}
        </div>
      </div>
      )}
      {!tasks && (
      <LoadingPanel />
      )}
    </div>
  );
}
