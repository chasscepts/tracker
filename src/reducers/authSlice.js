import { createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

/* eslint-disable no-param-reassign */
const authSlice = createSlice({
  name: 'auth',
  initialState: {
    authenticated: false,
    loginError: null,
    registrationError: null,
    token: null,
    registrationSuccess: false,
  },
  reducers: {
    setToken: (state, { payload }) => {
      state.authenticated = true;
      state.token = payload;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
    setLoginError: (state, { payload }) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = payload;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
    setRegistrationError: (state, { payload }) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = payload;
      state.registrationSuccess = false;
    },
    setRegistrationSuccess: (state) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = true;
    },
    logout: (state) => {
      state.authenticated = false;
      state.token = null;
      state.loginError = null;
      state.registrationError = null;
      state.registrationSuccess = false;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  setToken,
  setLoginError,
  setRegistrationError,
  setRegistrationSuccess,
  logout,
} = authSlice.actions;

export const login = (email, password) => (dispatch) => {
  const url = 'http://localhost/auth';
  axios({
    url,
    method: 'post',
    responseType: 'json',
    data: { email, password },
  })
    .then((res) => {
      dispatch(setToken(res.data.token));
    })
    .catch((err) => {
      if (!err) {
        dispatch(setLoginError('Unknown error encountered. Please try again.'));
        return;
      }
      if (err.response) {
        dispatch(setLoginError(err.response.data));
        return;
      }
      if (err.request) {
        dispatch(setLoginError('Server is not responding.'));
        return;
      }
      if (err.message) {
        dispatch(setLoginError(err.message));
        return;
      }
      if (typeof err === 'string') {
        dispatch(setLoginError(err));
        return;
      }
      dispatch(setLoginError('Unknown error encountered. Please try again.'));
    })
    .catch((err) => {
      console.log(err);
    });
};

export const register = (email, password) => (dispatch) => {
  const url = 'http://localhost/auth/register';
  axios({
    url,
    method: 'post',
    responseType: 'json',
    data: { email, password },
  })
    .then(() => {
      dispatch(setRegistrationSuccess());
    })
    .catch((err) => {
      if (!err) {
        dispatch(setRegistrationError('An unknown error encountered. Please try again.'));
        return;
      }
      if (err.response) {
        dispatch(setRegistrationError(err.response.data));
        return;
      }
      if (err.request) {
        dispatch(setRegistrationError('Server is not responding.'));
        return;
      }
      if (err.message) {
        dispatch(setRegistrationError(err.message));
        return;
      }
      if (typeof err === 'string') {
        dispatch(setRegistrationError(err));
        return;
      }
      dispatch(setRegistrationError('An unknown error encountered. Please try again.'));
    })
    .catch((err) => {
      console.log(err);
    });
};

export const selectAuthenticated = (state) => state.auth.authenticated;

export const selectLoginError = (state) => state.auth.loginError;

export const selectRegistrationError = (state) => state.auth.registrationError;

export default authSlice.reducer;
