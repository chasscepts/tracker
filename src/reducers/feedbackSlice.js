import { createSlice } from '@reduxjs/toolkit';

/* eslint-disable no-param-reassign */
const feedbackSlice = createSlice({
  name: 'feedback',
  initialState: {
    id: 0,
    feedbacks: [],
  },
  reducers: {
    addFeedback: (state, { payload }) => {
      state.feedbacks.push({ id: state.id, payload });
      state.id += 1;
    },
    removeFeedback: (state, { payload }) => {
      state.feedbacks = state.feedbacks.filter((f) => f.id !== payload);
    },
  },
});
/* eslint-enable no-param-reassign */

export const { addFeedback, removeFeedback } = feedbackSlice.actions;

export const selectFeedbacks = (state) => state.feedback.feedbacks;

export default feedbackSlice.reducer;
