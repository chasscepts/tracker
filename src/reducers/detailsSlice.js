import { createSlice } from '@reduxjs/toolkit';
import { fetchGroupTasks } from '../api';
import { addFeedback } from './feedbackSlice';

/* eslint-disable no-param-reassign */
const detailsSlice = createSlice({
  name: 'details',
  initialState: {
    requestCount: 0,
    hasPendingRequest: false,
    store: [],
    group: null,
  },
  reducers: {
    addRequest: (state) => {
      state.requestCount += 1;
      state.hasPendingRequest = true;
    },
    removeRequest: (state) => {
      let temp = state.requestCount - 1;
      if (temp < 0) {
        temp = 0;
        state.hasPendingRequest = false;
      }
      state.requestCount = temp;
    },
    addGroup: (state, { payload }) => {
      state.group = payload;
      state.store.push(payload);
      let temp = state.requestCount - 1;
      if (temp < 0) {
        temp = 0;
        state.hasPendingRequest = false;
      }
      state.requestCount = temp;
    },
    setGroup: (state, { payload }) => {
      state.group = payload;
      let temp = state.requestCount - 1;
      if (temp < 0) {
        temp = 0;
        state.hasPendingRequest = false;
      }
      state.requestCount = temp;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  addRequest,
  removeRequest,
  addGroup,
  setGroup,
} = detailsSlice.actions;

export const selectDetailsHasPendingError = (state) => state.details.hasPendingRequest;
export const selectDetailsGroup = (state) => state.details.group;

export const loadGroupTasks = (groupId, title) => (dispatch, getState) => {
  const { auth: { user }, details: { store } } = getState();
  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to fetch tasks', type: 'error' }));
    return;
  }

  const group = store.find((g) => g.id === groupId);
  if (group) {
    dispatch(setGroup(group));
    return;
  }

  dispatch(addRequest());

  fetchGroupTasks(user.token, groupId)
    .then((group) => {
      dispatch(addGroup({ title, id: groupId, tasks: group }));
    })
    .catch(({ message }) => {
      dispatch(removeRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export default detailsSlice.reducer;
