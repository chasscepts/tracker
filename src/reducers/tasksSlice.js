import { createSlice } from '@reduxjs/toolkit';
import {
  fetchGroups, fetchTasks, createTask, updateTask, deleteTask,
} from '../api';
import { today, daysAgo } from '../utilities/dates';
import { addFeedback } from './feedbackSlice';

/* eslint-disable no-param-reassign */
const tasksSlice = createSlice({
  name: 'tasks',
  initialState: {
    date: today(),
    groups: null,
    groupsLoading: false,
    tasksCache: {},
    tasks: [],
    tasksLoading: false,
    createRequestCount: 0,
    updateRequestCount: 0,
  },
  reducers: {
    setDate: (state, { payload }) => {
      state.date = payload;
    },
    setGroups: (state, { payload }) => {
      state.groups = payload;
      state.groupsLoading = false;
    },
    setGroupsLoading: (state, { payload }) => {
      state.groupsLoading = payload;
    },
    setTasks: (state, { payload }) => {
      state.tasks = payload;
    },
    setTasksLoading: (state, { payload }) => {
      state.tasksLoading = payload;
    },
    updateTasksCache: (state, { payload: { date, tasks } }) => {
      state.tasksCache[date] = tasks;
    },
    addCreateTaskRequest: (state) => {
      state.createRequestCount += 1;
    },
    addTask: (state, { payload }) => {
      state.tasks.push(payload);
      if (state.createRequestCount > 0) {
        state.createRequestCount -= 1;
      }
    },
    removeCreateRequest: (state) => {
      if (state.createRequestCount > 0) {
        state.createRequestCount -= 1;
      }
    },
    addUpdateTaskRequest: (state) => {
      state.updateRequestCount += 1;
    },
    removeUpdateTaskRequest: (state) => {
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    updateLocalTask: (state, { payload: { id, title } }) => {
      const task = state.tasks.find((t) => t.id === id);
      if (task) {
        task.title = title;
      }
      Object.keys(state.tasksCache).forEach((key) => {
        const task = state.tasksCache[key].find((t) => t.id === id);
        if (task) {
          task.title = title;
        }
      });
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    deleteLocalTask: (state, { payload }) => {
      state.tasks = state.tasks.filter((t) => t.id !== payload);
      if (state.updateRequestCount > 0) {
        state.updateRequestCount -= 1;
      }
    },
    updateTaskEntry: (state, { payload: { entry, duration } }) => {
      const task = state.tasks.find((t) => t.id === entry.task_id);
      if (task) {
        task.entries[0].duration += duration;
      }
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  setDate,
  setGroups,
  setGroupsLoading,
  setTasks,
  setTasksLoading,
  updateTasksCache,
  addTask,
  addCreateTaskRequest,
  removeCreateRequest,
  updateLocalTask,
  addUpdateTaskRequest,
  removeUpdateTaskRequest,
  deleteLocalTask,
  updateTaskEntry,
} = tasksSlice.actions;

export const selectDate = (state) => state.tasks.date;

export const selectGroups = (state) => state.tasks.groups;

export const selectGroupsLoading = (state) => state.tasks.groupsLoading;

export const selectTasks = (state) => state.tasks.tasks;

export const selectCreateTaskRequestCount = (state) => state.tasks.createRequestCount;

export const selectUpdateTaskRequestCount = (state) => state.tasks.updateRequestCount;

export const loadGroups = () => (dispatch, getState) => {
  const state = getState();
  if (state.tasks.groups || state.tasks.groupsLoading) return;

  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to fetch groups', type: 'error' }));
    return;
  }

  dispatch(setGroupsLoading(true));

  fetchGroups(user.token)
    .then((groups) => {
      dispatch(setGroups(groups));
    })
    .catch(({ message }) => {
      dispatch(setGroupsLoading(false));
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const loadTasks = () => (dispatch, getState) => {
  const state = getState();

  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to fetch tasks!', type: 'error' }));
    return;
  }

  dispatch(setTasksLoading());

  fetchTasks(user.token, { start: daysAgo(14) })
    .then((tasks) => {
      dispatch(setTasks(tasks));
    })
    .catch(({ message }) => dispatch(addFeedback({ message, type: 'error' })));
};

export const createTaskAsync = (groupId, title) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to create tasks', type: 'error' }));
    return;
  }

  dispatch(addCreateTaskRequest());

  createTask(user.token, groupId, title)
    .then((task) => {
      dispatch(addTask(task));
      dispatch(addFeedback({ message: `The task ${task.title} was successfully created.`, type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeCreateRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const updateTaskAsync = (id, title) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to update tasks', type: 'error' }));
    return;
  }

  dispatch(addUpdateTaskRequest());

  updateTask(user.token, id, title)
    .then(() => {
      dispatch(updateLocalTask({ id, title }));
      dispatch(addFeedback({ message: `Task ${title} successfully updated.`, type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeUpdateTaskRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export const deleteTaskAsync = (id) => (dispatch, getState) => {
  const state = getState();
  const { auth: { user } } = state;

  if (!user) {
    dispatch(addFeedback({ message: 'You must be logged in to update tasks', type: 'error' }));
    return;
  }

  dispatch(addUpdateTaskRequest());

  deleteTask(user.token, id)
    .then(() => {
      dispatch(deleteLocalTask(id));
      dispatch(addFeedback({ message: 'Task successfully deleted.', type: 'success' }));
    })
    .catch(({ message }) => {
      dispatch(removeUpdateTaskRequest());
      dispatch(addFeedback({ message, type: 'error' }));
    });
};

export default tasksSlice.reducer;
