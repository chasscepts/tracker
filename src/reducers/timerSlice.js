import { createSlice } from '@reduxjs/toolkit';
import api from '../api';
import { setTasksError } from './tasksSlice';

/* eslint-disable no-param-reassign */
const timerSlice = createSlice({
  name: 'timer',
  initialState: {
    entry: {},
    nextEntry: {},
    updateError: null,
    updateSuccess: null,
    pendingRequestCount: 0,
    hasPendingRequests: false,
  },
  reducers: {
    commitNext: (state) => {
      state.entry = state.nextEntry;
      state.nextEntry = null;
    },
    setEntry: (state, { payload }) => {
      state.entry = payload;
    },
    setNextEntry: (state, { payload }) => {
      if (!state.entry.id) {
        state.entry = payload;
      } else {
        state.nextEntry = payload;
      }
    },
    addEntryRequest: (state) => {
      state.pendingRequestCount += 1;
      state.hasPendingRequests = true;
      state.updateError = null;
      state.updateSuccess = null;
    },
    updateStoreEntry: (state, { payload: { message, duration } }) => {
      console.log({ message, duration });
      state.updateSuccess = message;
      state.entry.entry.duration += duration;
      console.log(state.entry.entry.duration);
      let temp = state.pendingRequestCount - 1;
      if (temp <= 0) {
        state.hasPendingRequests = false;
        temp = 0;
      }
      state.pendingRequestCount = temp;
    },
    setEntryUpdateSuccess: (state, { payload }) => {
      state.updateSuccess = payload;
      let temp = state.pendingRequestCount - 1;
      if (temp <= 0) {
        state.hasPendingRequests = false;
        temp = 0;
      }
      state.pendingRequestCount = temp;
    },
    setEntryUpdateError: (state, { payload }) => {
      state.updateError = payload;
      let temp = state.pendingRequestCount - 1;
      if (temp <= 0) {
        state.hasPendingRequests = false;
        temp = 0;
      }
      state.pendingRequestCount = temp;
    },
  },
});
/* eslint-enable no-param-reassign */

export const {
  commitNext,
  setEntry,
  setNextEntry,
  addEntryRequest,
  updateStoreEntry,
  setEntryUpdateSuccess,
  setEntryUpdateError,
} = timerSlice.actions;

export const selectEntry = (state) => state.timer.entry;
export const selectNextEntry = (state) => state.timer.nextEntry;
export const selectEntryUpdateError = (state) => state.timer.updateError;
export const selectEntryUpdateSuccess = (state) => state.timer.updateSuccess;
export const selectEntryHasPendingRequest = (state) => state.timer.hasPendingRequests;
export const selectEntryPendingRequestCount = (state) => state.timer.pendingRequestCount;

export default timerSlice.reducer;

export const updateEntry = (entry, duration, title) => (dispatch, getState) => {
  const { auth: { user } } = getState();

  if (!user) {
    dispatch(setTasksError('You must be logged in to fetch tasks'));
    return;
  }

  dispatch(addEntryRequest());

  api.updateEntry(user.token, entry, duration)
    .then(() => {
      const msg = `Update of Task Entry ${title} successful`;
      dispatch(updateStoreEntry({ message: msg, duration }));
    })
    .catch(({ message }) => {
      const msg = `Update of Task Entry ${title} failed with message: ${message}`;
      dispatch(setEntryUpdateError(msg));
    });
};
