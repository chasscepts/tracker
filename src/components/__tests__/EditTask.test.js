import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Wrapper from '../../test_helpers/wrapper';
import EditTasks from '../EditTasks';
import { updateTask, deleteTask } from '../../api';
import { createStore } from '../../app/store';
import login from '../../test_helpers/login';
import { setTasks } from '../../reducers/tasksSlice';

jest.mock('../../api');

const tasks = [
  {
    id: 1,
    title: 'Task 1',
    group_id: 1,
    user_id: 1,
    entries: [
      {
        id: 1, task_id: 1, entry_date: '2021-09-21', duration: 0,
      },
    ],
  },
];

describe('EditTask', () => {
  it('matches snapshot', async () => {
    const store = createStore();
    store.dispatch(setTasks(tasks));
    const tree = renderer
      .create(<Wrapper store={store} Component={EditTasks} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('update calls api with correct parameter', async () => {
    const store = createStore();
    login(store);
    store.dispatch(setTasks(tasks));
    const title = 'Updated Title';
    const task = tasks[0];

    updateTask.mockResolvedValue({ id: 1, title });

    render(<Wrapper store={store} Component={EditTasks} />);
    const input = screen.getByPlaceholderText('Enter New Title');
    userEvent.click(screen.getByRole('button', { name: `${task.title}` }));
    userEvent.clear(input);
    userEvent.type(input, title);
    userEvent.click(screen.getByRole('button', { name: 'Edit' }));
    await Promise.resolve();
    expect(updateTask).toHaveBeenCalledWith(expect.any(String), task.id, title);
  });

  it('delete calls api with correct paremeters', async () => {
    const store = createStore();
    login(store);
    store.dispatch(setTasks(tasks));
    const task = tasks[0];

    deleteTask.mockResolvedValue('');

    render(<Wrapper store={store} Component={EditTasks} />);
    userEvent.click(screen.getByRole('button', { name: `${task.title}` }));
    userEvent.click(screen.getByRole('button', { name: 'Delete' }));
    expect(deleteTask).toHaveBeenCalledWith(expect.any(String), task.id);
  });
});
