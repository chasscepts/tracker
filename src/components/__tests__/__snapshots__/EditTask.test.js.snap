// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTask matches snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "overflow": "hidden",
    }
  }
>
  <form
    onSubmit={[Function]}
    style={
      Object {
        "border": "1px solid #ddd",
        "borderRadius": "5px",
        "margin": "5px auto 15px",
        "maxWidth": "400px",
        "padding": "15px",
        "width": "100%",
      }
    }
  >
    <h2
      style={
        Object {
          "borderBottom": "1px dotted #62b5e5",
          "color": "#62b5e5",
          "marginBottom": "15px",
          "paddingBottom": "5px",
        }
      }
    >
      Edit Task
    </h2>
    <div
      style={
        Object {
          "color": "#9e0606",
          "fontSize": "0.8rem",
          "fontWeight": "bold",
          "marginBottom": "5px",
        }
      }
    >
      You have not selected any task
    </div>
    <input
      name="title"
      onChange={[Function]}
      placeholder="Enter New Title"
      style={
        Object {
          "border": "1px solid #ddd",
          "color": "inherit",
          "display": "block",
          "marginBottom": "15px",
          "padding": "7px",
          "width": "100%",
        }
      }
      value=""
    />
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "flex-end",
        }
      }
    >
      <button
        className="btn blue round"
        type="submit"
      >
        Edit
      </button>
      <button
        className="btn red round"
        onClick={[Function]}
        style={
          Object {
            "marginLeft": "20px",
          }
        }
        type="button"
      >
        Delete
      </button>
    </div>
  </form>
  <div
    style={
      Object {
        "flex": 1,
        "overflow": "auto",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexFlow": "row wrap",
          "justifyContent": "space-between",
          "padding": "10px",
        }
      }
    >
      <button
        name={1}
        onClick={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "border": "1px solid #ddd",
            "borderRadius": "4px",
            "color": "inherit",
            "cursor": "pointer",
            "display": "flex",
            "fontSize": "0.9rem",
            "fontWeight": "bold",
            "height": "80px",
            "justifyContent": "center",
            "margin": "10px",
            "padding": "5px",
            "width": "80px",
          }
        }
        type="button"
      >
        <span
          style={
            Object {
              "pointerEvents": "none",
            }
          }
        >
          Task 1
        </span>
      </button>
    </div>
  </div>
</div>
`;
